packages/angular_devkit/core/src/jobs/architecture.md

```diff
@@ -0,0 +1,253 @@
+    const i = setInterval(() => {
+      t++;
+      subscriber.next({ kind: JobEventKind.Output, value: t });
+      if (t === x) {
+        subscriber.next({ kind: JobEventKind.End });
+        subscriber.complete();
+      }
+    }, 1000);
+
+    return () => {
+      clearInterval(i);
+    };
+  })
+};
+
+// For now, jobs can not be called without a registry and scheduler.
+const registry = new SimpleJobRegistry();
+registry.register('output-from-1-to-x', output1ToXEverySecond, {
+  argument: { type: 'number' },
+  output: { type: 'number' },
+});
+const scheduler = new SimpleScheduler(registry);
+
+// Need to keep the same name that the registry would understand.
+// Count from 1 to 10.
+const job = scheduler.schedule('output-from-1-to-x', 10);
+
+// A Job<> instance has more members, but we only want the output values here.
+job.output.subscribe(x => console.log(x));
+```
+
+This seems like a lot of boilerplate in comparison, but there are a few advantages;
+
+1. lifecycle. Jobs can tell when they start doing work and when work is done.
+1. everything is typed, even at runtime.
+1. the context also contains an input Observable that receives typed input events, including
+    input values, and stop requests.
+1. jobs can also schedule other jobs and wait for them, even if they don't know if a job is
+    implemented in the system.
+
+## Diagram
+A simpler way to think about jobs in contrast to observables is that job are closer to a Unix
+process. It has an argument (command line flags), receive inputs (STDIN and interupt signals),
+and output values (STDOUT) as well as diagnostic (STDERR). They can be plugged one into another
+(piping), and can be transformed, synchronized and scheduled (fork, exec, cron).
+
+```plain
+- given A the type of the argument
+- given I the type of the input
+- given O the type of the output
+
+                              ,______________________   ,----> rawOuput: JobEvent<O>
+    rawInput: JobInput<I> --> | handler(argument: A) |            - JobEventKind.Log
+                              `⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻             - JobEventKind.Output
+                                                                  - ...
+```
+
+`JobInput` includes:
+
+1. `JobInputKind.Ping`. A simple message that should be answered with `JobEventKind.Pong` when
+    the job is responsive. The `id` field of the message should be used when returning `Pong`.
+1. `JobInputKind.Stop`. The job should be stopped. This is used when cancelling/unsubscribing from
+    the `output` (or by calling `stop()`). Any inputs or outputs after this event will be ignored.
+1. `JobInputKind.Input` is used when sending inputs to a job. These correspond to the `next`
+    methods of an `Observer` and are reported to the job through its `context.input` Observable.
+    There is no way to communicate an error to the job.
+
+`JobEvent` includes:
+
+1. `JobEventKind.Create`. The `Job<>` was created, its dependencies are done, and the library is 
```

Comment:
This is `JobEventKind.Ready` now, right?

Reply:
Done.